src/visitors/semantic_analysis_visitor.cpp

```diff
@@ -71,27 +73,56 @@ void SemanticAnalysisVisitor::visit_program(const ast::Program& node) {
     node.visit_children(*this);
 }
 
+void SemanticAnalysisVisitor::visit_function_call(const ast::FunctionCall& node) {
+    /// <-- This code is for check 9
+    if (visited_function_or_procedure_blocks.empty()) {
+        return;
+    }
+    const auto func_name = node.get_node_name();
+    const auto func_already_visited = std::find_if(visited_function_or_procedure_blocks.begin(),
+                                                   visited_function_or_procedure_blocks.end(),
+                                                   [&func_name](const ast::Block* block) {
+                                                       return func_name == block->get_node_name();
+                                                   });
+    if (func_already_visited != visited_function_or_procedure_blocks.end()) {
+        logger->critical(
+            fmt::format("SemanticAnalysisVisitor :: Recursive function call of \"{}\" in {}",
+                        func_name,
+                        node.get_token()->position()));
+        check_fail = true;
+        return;
+    }
+    auto func_symbol = psymtab->lookup(func_name);
```

Comment:
is there a big penalty to have this just be
```suggestion
    auto func_symbol = node.get_symbol_table()->lookup(func_name);
```
and get around having a member pointer to the program table in the visitor class?

Reply:
Here we need the global symbol table that has all the symbols in the mod file to be able to search whether there is a `FUNCTION` or `PROCEDURE` defined with the `func_name` so we need to set `psymtab` when we visit the `ast::Program` which is the top level symbol table